### ISSUE TITLE
Building with scons > 4.0 results in an empty compile_commands.json

### ISSUE THREAD

**Issue by object71** [godotengine/godot#54434]
### Godot version

4.0.dev (727c51f35d)

### System information

Windows 10

### Issue description

When compiling using:

- python version ```3.9.7```
- scons version of ```4.2.0```
- the following command: ```scons platform=windows tools=yes -j 12```

... I get a compile_commands.json file in the root project directory but the file itself is only containing empty array brackets:
```json
[]
```

This is undesirable for me as I like to use CLion as an IDE for the project and it can handle compile_commands as long as they exist. It doesn't help to also specify the output file as mentioned in #39883 and will still produce nothing.

### Steps to reproduce

1. Clone the repository
2. Run ```scons platform=windows compile_commands.json```
3. Observe compile_commands.json only containing an empty array

**Comment by object71** [comment]
A workaround is to use [scan-build](https://github.com/rizsotto/scan-build) or [bear](https://github.com/rizsotto/Bear)

```bash
pip install scan-build
```

...and then...

```bash
intercept-build scons <options>
```

**Comment by akien-mga** [comment]
It works fine on Linux, so might be Windows/MSVC specific.

**Comment by fire** [comment]
This is known for the msvc compiler. May be documented.

**Comment by viktor-ferenczi** [comment]
Attempted to fix this, because I need a compilation database on Windows to hook up CLion for Godot development.

Looked into the source code of SCons again. I fixed the random crashes of parallel build on Windows back in 2018, that was the first time.

SCons hasn't improved at all since then. Complicated spaghetti code, bad use of Python everywhere. It is unnecessarily complex. Due to bad naming it hard to understand what is going on and fix simple issues like this. 

The use of SCons may eventually end up as a big time sink. I would suggest switching to another build tool in the long term. Some research needs to be done which one.

See [some further rant](https://news.ycombinator.com/item?id=20140270) here.

Trying to use workarounds. It is ugly.

**Comment by viktor-ferenczi** [comment]
@Calinou Thanks for the additional information. Good that we have a better alternative in the pipeline.

**Comment by viktor-ferenczi** [comment]
On Windows the workaround is to add the `vsproj=yes` option to your SCons build command. [It will create a Visual Studio solution](https://docs.godotengine.org/en/latest/development/cpp/configuring_an_ide/visual_studio.html), which can be opened with CLion as well.

**Comment by viktor-ferenczi** [comment]
I suggest closing this ticket, since we have easy workarounds on all platforms and #51153 will replace SCons in the long term.

**Comment by akien-mga** [comment]
Please leave rants and buildsystem advocacy out of this issue, it's off topic.

@dmoody256 Do you know if SCons' `compilation_db` tool is meant to support MSVC? IIRC support was missing in the initial implementation but I don't find any upstream issue tracking that.

**Comment by dmoody256** [comment]
@akien-mga There are tests which test basic use, and we run them with MSVC and they pass, so yeah its supports MSVC.

I went a head checked out the commit in the description, ran compiledb on windows 10 with python 3.9 and scons 4.2. The problem reproduced for me when my system was not configured correctly (I didn't have MSVC installed correctly). Also the build was always selecting MingW, even when I set msvc_version. I had to make this change to make it use MSVC:
```
D:\godot>git diff
diff --git a/platform/windows/detect.py b/platform/windows/detect.py
index 6752885f37..10ae00b6bb 100644
--- a/platform/windows/detect.py
+++ b/platform/windows/detect.py
@@ -479,7 +479,7 @@ def configure(env):
         setup_msvc_manual(env)
         env.msvc = True
         manual_msvc_config = True
-    elif env.get("MSVC_VERSION", "") and not env["use_mingw"]:
+    elif env.get("msvc_version", "") and not env["use_mingw"]:
         setup_msvc_auto(env)
         env.msvc = True
         manual_msvc_config = False
```

After I fixed that and fixed my Visual Studio  2019 installation, was able to generate good compile_commands.json files with the following command:
```
scons platform=windows tools=yes -j 12 compile_commands.json msvc_version=14.2
```

**Comment by akien-mga** [comment]
Awesome, thanks for checking! CC @bruvzg to review the patch and the MSVC detection issue.

**Comment by bruvzg** [comment]
I do not think the patch is correct, IIRC MSVC_VERSION is internal Scons variable used to indicate it has auto-detected MSVC.

Probably the option should be changed to upper case instead, or we should check both. But I can't test it right now.

**Comment by dmoody256** [comment]
the `setup_msvc_auto` will set the scons internal `MSVC_VERSION` from the godot `msvc_version`

**Comment by bruvzg** [comment]
> the setup_msvc_auto function will set the scons internal MSVC_VERSION from the godot msvc_version

Oh, I have not noticed it, it seems to do it indeed. Then the patch should be fine.

**Comment by viktor-ferenczi** [comment]
@dmoody256 Applied the same patch to the Godot 4.0 master. It does not work for me. 
- The build still generates an empty list in the JSON file, even if I do a full build from scratch.
- Tried to reinstall VS2019 to make sure MSVC is installed correctly. 
- Also reinstalled SCons to have the latest version for sure.

The build command was:
`scons -j 12 platform=windows target=release_debug tools=yes compiledb=yes msvc_version=14.2`

Appending `compile_commands.json` at the end and/or removing `target=release_debug` gives the same result.

Removing `compiledb=yes` gives: `Do not know how to make File target 'compile_commands.json'`

I've re-read the relevant documentation, still no idea what I do wrong.

Do I overlook something?

**Comment by dmoody256** [comment]
@viktor-ferenczi I did some more testing and found when I run the build in a shell environment without MSVC setup, the compile_commands.json is also empty. If you run the developer tools scripts to setup the current shell does it start working, for example you could run:
```
"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat"
```

SCons internally should setup this for any spawned shells it runs commands in, but it looks like compilation_db tool is not taking that into account.

**Comment by viktor-ferenczi** [comment]
@dmoody256 Thank you, adding a
`CALL "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat"`
to my build batch file right before the `scons ...` line helped. Now it produces a compilation database as expected.

I think this ticket ends up [fixing the documentation](https://docs.godotengine.org/en/latest/search.html?q=%22compiledb%3Dyes%22&check_keywords=yes&area=default#) then.

These two places:
- [CLion](https://docs.godotengine.org/en/latest/development/cpp/configuring_an_ide/clion.html#importing-the-project)
- [VSCode](https://docs.godotengine.org/en/latest/development/cpp/configuring_an_ide/visual_studio_code.html#importing-the-project)

I guess we should describe how to create a compilation database at one place, then refer to that one from both pages.

**Comment by dmoody256** [comment]
I looked into it more and the issue appears to be that the compilation_db tool is initialized to early in the godot SConstruct. The compilation_db tool hooks itself into the current compilation related Builder objects that exist in the current environment. The compilation_db tool should be initialized later in the SConstruct, to make sure it hooks into any builders which may have been added when configuring and before calling said builders (e.g. calling the SCsub tree).

Here is a patch which moves the compilation_db init later so that it will include the msvc compilation builders without needing to setup the environment:
```
D:\godot>git diff
diff --git a/SConstruct b/SConstruct
index 94bc53eaa5..700b482d81 100644
--- a/SConstruct
+++ b/SConstruct
@@ -344,10 +344,6 @@ if selected_platform in platform_list:

     scons_ver = env._get_major_minor_revision(scons_raw_version)

-    if scons_ver >= (4, 0, 0):
-        env.Tool("compilation_db")
-        env.Alias("compiledb", env.CompilationDatabase())
-
     # 'dev' and 'production' are aliases to set default options if they haven't been set
     # manually by the user.
     if env["dev"]:
@@ -715,6 +711,10 @@ if selected_platform in platform_list:
         env.vs_incs = []
         env.vs_srcs = []

+    if scons_ver >= (4, 0, 0):
+        env.Tool("compilation_db")
+        env.Alias("compiledb", env.CompilationDatabase())
+
     Export("env")

     # Build subdirs, the build order is dependent on link order.
diff --git a/platform/windows/detect.py b/platform/windows/detect.py
index 6752885f37..10ae00b6bb 100644
--- a/platform/windows/detect.py
+++ b/platform/windows/detect.py
@@ -479,7 +479,7 @@ def configure(env):
         setup_msvc_manual(env)
         env.msvc = True
         manual_msvc_config = True
-    elif env.get("MSVC_VERSION", "") and not env["use_mingw"]:
+    elif env.get("msvc_version", "") and not env["use_mingw"]:
         setup_msvc_auto(env)
         env.msvc = True
         manual_msvc_config = False
```

**Comment by akien-mga** [comment]
Thanks a lot! I made a PR with that fix, moving the whole `compiledb` block: #65440.
(We make it opt-in to save on incremental rebuild time, saves around 3 s last I checked: #57820.)

**Comment by bruvzg** [comment]
> Then the patch should be fine.

Actually, it's probably still not OK, since it will not work if msvc_version option is not set, and internal variable is set (in case msvc auto detection).

So this check still have to check both.

I would also rename the option to avoid confusion.
